### PR TITLE
Switch the intrinsic names to a string table

### PR DESCRIPTION
**Pull request by chandlerc**
This avoids the need to dynamically relocate each pointer in the table.

To make this work, this PR also moves the binary search of intrinsic names to an internal function with an adjusted signature, and switches the unittesting to test against actual intrinsics.

### CODE REVIEW

File: llvm/lib/IR/Intrinsics.cpp [hunk 1]
```diff
@@ -638,37 +638,53 @@ int llvm::Intrinsic::lookupLLVMIntrinsicByName(ArrayRef<const char *> NameTable,
...
+  const int *Low = NameOffsetTable.begin();
+  const int *High = NameOffsetTable.end();
+  const int *LastLow = Low;
   while (CmpEnd < Name.size() && High - Low > 0) {
     size_t CmpStart = CmpEnd;
     CmpEnd = Name.find('.', CmpStart + 1);
     CmpEnd = CmpEnd == StringRef::npos ? Name.size() : CmpEnd;
-    auto Cmp = [CmpStart, CmpEnd](const char *LHS, const char *RHS) {
-      return strncmp(LHS + CmpStart, RHS + CmpStart, CmpEnd - CmpStart) < 0;
+    auto Cmp = [CmpStart, CmpEnd](auto LHS, auto RHS) {
```
Comment (jurahul) [inline llvm/lib/IR/Intrinsics.cpp:648]: Can this be just int now? I assume the comparator is now used with const char * types directly anymore.

**Review by jurahul — COMMENTED** [review]
Just one comment above, LGTM otherwise! Thanks for finishing this, I had started on this but got sidetracked.

File: llvm/utils/TableGen/IntrinsicEmitter.cpp [hunk 2]
```diff
@@ -239,13 +239,37 @@ static constexpr IntrinsicTargetInfo TargetInfos[] = {
...
 #ifdef GET_INTRINSIC_NAME_TABLE
 // Note that entry #0 is the invalid intrinsic!
+
+)";
+
+  Table.EmitStringLiteralDef(OS, "static constexpr char IntrinsicNameTable[]",
+                             /*Indent=*/"");
+
+  OS << R"(
+static constexpr int IntrinsicNameOffsetTable[] = {
```
Comment (nikic) [inline llvm/utils/TableGen/IntrinsicEmitter.cpp:259]:
```suggestion
static constexpr unsigned IntrinsicNameOffsetTable[] = {
```
Or `uint32_t` -- or can these offset be negative?

Comment (chandlerc) [reply llvm/utils/TableGen/IntrinsicEmitter.cpp]: 🤷🏻 I just default to `int` when it doesn't matter. But I guess LLVM uses `unsigned` more widely, so sure.

**Comment by jurahul** [comment]
They cannot be negative.

**Review by chandlerc — COMMENTED** [review]
Switched to `unsigned`, PTAL.

File: llvm/lib/IR/Intrinsics.cpp (hunk 1 above)
Comment (chandlerc) [reply llvm/lib/IR/Intrinsics.cpp:648]: No, sadly. We are searching for a string within the offset-strings, so we need to handle both `const char*` and `unsigned`.

And `std::equal_range` requires this to work in either direction.I tried to leave a comment on the next line about this.

**Review by jurahul — APPROVED** [review]
LGTM to me. Approving.

**Comment by chandlerc** [comment]
Thanks for review, merging once pre-merge builds come back.

**Pull request merged**